go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb

```diff
@@ -180,7 +159,16 @@ def run_go_get
 
           File.write(tmp_go_file, "package dummypkg\n") unless package
 
-          _, stderr, status = Open3.capture3(ENVIRONMENT, "go get -d")
+          # TODO: go 1.18 will make `-d` the default behavior, so remove the flag then
+          command = +"go get -d"
+          # `go get` accepts multiple packages, each separated by a space
+          dependencies.each do |dep|
+            # Use version pinning rather than `latest` just in case
+            # a new version gets released in the middle of our run.
+            version = "v" + dep.version.sub(/^v/i, "")
+            command << " #{dep.name}@#{version}"
+          end
+          _, stderr, status = Open3.capture3(ENVIRONMENT, command)
```

Comment:
`spec/dependabot/go_modules/file_updater/go_mod_updater_spec.rb:245` is failing because we now no longer run `go get -d` on all the modules, so basically we're only running the code for `update_go_mod` (but now using `go get instead of our custom helper), we still need to run a bare `go get -d` to verify that the packages we have downloaded have the module that we expect to find etc.

I verified the tests pass after adding a `go get -d` after this command, but what might be nice is keeping the original `def run_go_get` method, and moving the contents of this method into `def update_go_mod`

I think the tests should pass after that

Reply:
Ah, so that was intentional... 

I'd thought the generic `go get -d` wasn't needed because we now rely on `go get lib@version` to _be smart enough to know how to upgrade a particular library_ and if some other part of the dep tree is hosed, then we let `go get -d lib@version` decide whether that is relevant or not. I don't expect (or want) dependabot to validate my _existing_ dep tree, I only expect it to validate that version bumps don't put the dep tree in a worse state.

I still think that actually. 

So let's do this, I'll break this PR in two. The first will keep this generic `go get -d` command, and simply swap out the custom helper for `go get` and then the second will be a discussion of whether this generic command is within the scope of what Dependabot should be doing. 

That should keep the majority of this PR in a simple, non-controversial PR, and then the second PR (which will be more controversial) can be much more easily reasoned about/researched/discussed. 

I'm also reasonably certain this first PR will _not_ solve #3526 which was my original impetus for all this work. That'd require the second PR.

Reply:
Yeah I get what you're saying, however:

>  I only expect it to validate that version bumps don't put the dep tree in a worse state.

I think that running the generic `go get -d` still protects against this, for the case the failing test was validating, it prevents updating to a version when the package has been renamed, right?

Maybe it is an edge-case we should not worry about, assume peoples CI will fail and they'll figure out what the issue is and update the import statements etc manually, but I definitely appreciate you pulling these out into separate PRs so we can discuss more focussed 🙇

Reply:
Also, a few notes to myself when I get time to dig into this:
1. IIRC, the other reason this was originally here was because we'd manually munged `go.mod` w/o touching `go.sum`, so we needed this `go get` to pickup those `go.sum` changes. (This could have also been done via `go mod tidy`, but `go get` was handling it well enough). This is why I originally thought this was no longer needed.
2. The transitory resolving stuff will likely get cleaned up a lot more as part of [lazy loading in `go 1.17`](https://tip.golang.org/doc/go1.17#go-command).
3. There's still the open question though of should dependabot fail when something deep upstream fails... eg, this started failing yesterday 4 steps up our dep chain: https://github.com/hashicorp/go-discover/issues/172. I can see valid arguments both pro/con on this.